### PR TITLE
Parse Loops

### PR DESCRIPTION
**Pull request by chriselrod**
Closes #16

### CODE REVIEW

**Comment by chriselrod** [comment]
@sumiya11 If you're curious, we now have basic front end parsing!

```
> opt -mcpu=native --load-pass-plugin=/home/chriselrod/Documents/progwork/cxx/LoopModels/builddirnocov/libTurboLoop.so --passes=turbo-loop --disable-output /home/chriselrod/Documents/progwork/cxx/LoopPlayground/LoopInductTests/test/triangular_solve2.ll
# ...
orth opt succeeded!
optDeps.hasValue() = 1
LoopBlock schedule:mem->nodeIndex = 0; mem->ref = ar.indexMatrix() =
[  1  0
   0  1 ]
ArrayReference %24 (dim = 2, num loops: 2, element size: 8):
A.numRow() = 2; A.numCol() = 2
Sizes: [ unknown, %21 ]
Subscripts: [ i_1 , i_0 ]
s.getPhi() =
[  1  0
   0  1 ]
s.getOmega() = [ 0, 2, 0, 0, 0 ]

mem->nodeIndex = 0; mem->ref = ar.indexMatrix() =
[  1  0
   0  1 ]
ArrayReference %27 (dim = 2, num loops: 2, element size: 8):
A.numRow() = 2; A.numCol() = 2
Sizes: [ unknown, %9 ]
Subscripts: [ i_1 , i_0 ]
s.getPhi() =
[  1  0
   0  1 ]
s.getOmega() = [ 0, 2, 0, 0, 0 ]

mem->nodeIndex = 1; mem->ref = ar.indexMatrix() =
[  1  0
   0  1 ]
ArrayReference %27 (dim = 2, num loops: 2, element size: 8):
A.numRow() = 2; A.numCol() = 2
Sizes: [ unknown, %9 ]
Subscripts: [ i_1 , i_0 ]
s.getPhi() =
[  1  0
   0  1 ]
s.getOmega() = [ 0, 1, 0, 0, 0 ]

mem->nodeIndex = 1; mem->ref = ar.indexMatrix() =
[  1  1
   0  0 ]
ArrayReference %35 (dim = 2, num loops: 2, element size: 8):
A.numRow() = 2; A.numCol() = 2
Sizes: [ unknown, %32 ]
Subscripts: [ i_1 , i_1 ]
s.getPhi() =
[  1  0
   0  1 ]
s.getOmega() = [ 0, 1, 0, 0, 0 ]

mem->nodeIndex = 1; mem->ref = ar.indexMatrix() =
[  1  0
   0  1 ]
ArrayReference %27 (dim = 2, num loops: 2, element size: 8):
A.numRow() = 2; A.numCol() = 2
Sizes: [ unknown, %9 ]
Subscripts: [ i_1 , i_0 ]
s.getPhi() =
[  1  0
   0  1 ]
s.getOmega() = [ 0, 1, 0, 0, 0 ]

mem->nodeIndex = 1; mem->ref = ar.indexMatrix() =
[  1  0
   0  1 ]
ArrayReference %27 (dim = 2, num loops: 2, element size: 8):
A.numRow() = 2; A.numCol() = 2
Sizes: [ unknown, %9 ]
Subscripts: [ i_1  + 1, i_0 ]
s.getPhi() =
[  1  0
   0  1 ]
s.getOmega() = [ 0, 1, 0, 0, 0 ]

mem->nodeIndex = 1; mem->ref = ar.indexMatrix() =
[  1  1
   0  0 ]
ArrayReference %35 (dim = 2, num loops: 2, element size: 8):
A.numRow() = 2; A.numCol() = 2
Sizes: [ unknown, %32 ]
Subscripts: [ i_1  + 1, i_1 ]
s.getPhi() =
[  1  0
   0  1 ]
s.getOmega() = [ 0, 1, 0, 0, 0 ]

mem->nodeIndex = 1; mem->ref = ar.indexMatrix() =
[  1  0
   0  1 ]
ArrayReference %27 (dim = 2, num loops: 2, element size: 8):
A.numRow() = 2; A.numCol() = 2
Sizes: [ unknown, %9 ]
Subscripts: [ i_1  + 1, i_0 ]
s.getPhi() =
[  1  0
   0  1 ]
s.getOmega() = [ 0, 1, 0, 0, 0 ]

mem->nodeIndex = 2; mem->ref = ar.indexMatrix() =
[  0  0
   1  0
   0  1 ]
ArrayReference %27 (dim = 2, num loops: 3, element size: 8):
A.numRow() = 3; A.numCol() = 2
Sizes: [ unknown, %9 ]
Subscripts: [ i_1 , i_0 ]
s.getPhi() =
[  1  1  0
   0  0  1
   0  1  0 ]
s.getOmega() = [ 0, 0, 0, 0, 0, 0, 0 ]

mem->nodeIndex = 2; mem->ref = ar.indexMatrix() =
[  1  0
   1  0
   0  1 ]
ArrayReference %27 (dim = 2, num loops: 3, element size: 8):
A.numRow() = 3; A.numCol() = 2
Sizes: [ unknown, %9 ]
Subscripts: [ i_1  + i_2  + 2, i_0 ]
s.getPhi() =
[  1  1  0
   0  0  1
   0  1  0 ]
s.getOmega() = [ 0, 0, 0, 0, 0, 0, 0 ]

mem->nodeIndex = 2; mem->ref = ar.indexMatrix() =
[  1  0
   1  1
   0  0 ]
ArrayReference %35 (dim = 2, num loops: 3, element size: 8):
A.numRow() = 3; A.numCol() = 2
Sizes: [ unknown, %32 ]
Subscripts: [ i_1  + i_2  + 2, i_1 ]
s.getPhi() =
[  1  1  0
   0  0  1
   0  1  0 ]
s.getOmega() = [ 0, 0, 0, 0, 0, 0, 0 ]

mem->nodeIndex = 2; mem->ref = ar.indexMatrix() =
[  1  0
   1  0
   0  1 ]
ArrayReference %27 (dim = 2, num loops: 3, element size: 8):
A.numRow() = 3; A.numCol() = 2
Sizes: [ unknown, %9 ]
Subscripts: [ i_1  + i_2  + 2, i_0 ]
s.getPhi() =
[  1  1  0
   0  0  1
   0  1  0 ]
s.getOmega() = [ 0, 0, 0, 0, 0, 0, 0 ]
```
This orthogonalizes stores in the triangular solve example:
```julia
julia> function triangular_solve!(A, B, U)
           M, N = size(A)
           @inbounds for m = 1:M
               for n = 1:N
                   A[m,n] = B[m,n]
               end
               for k = 1:N
                   A[m,k] /= U[k,k]
                   for n = k+1:N
                       A[m,n] -= A[m,k]*U[k,n]
                   end
               end
           end
       end
```

Note that because we use `SCEV` for processing loop bounds and array indices, we read both as a function of iteration number. Thus, loop bounds effectively all start at zero.
I may update our Polyhedra representation to make this implicit, so we don't need to explicitly store the 0 lower bounds.
The Simplex class of course already has implicit 0 lower bounds, and supports a more efficient bounds-pruning algorithm.

In other words, the above loops with 1-based indexing are essentially translated to the following loops with 0-based indexing:
```julia
for m = 0:M-1
    for n = 0:N-1
        A[m,n] = B[m,n]
    end
    for k = 0:N-1
        A[m,k] /= U[k,k]
        for n = 0:N-k-2
            A[m,n+k+1] -= A[m,k] * U[k,n+k+1]
        end
    end
end
```
What the output I pasted above shows, is that we've rotated and fused all the loops, moving the reduction to the inner most loop.

I don't think the last omegas should be all zero though...
Also, I'll have to double check my math and code, because the output above says `n+k+2` and not `n+k+1` for indexing...
